materialize-mysql/driver.go

```diff
@@ -527,69 +540,105 @@ func (t *transactor) addBinding(ctx context.Context, target sql.Table, varchars
 	return nil
 }
 
-type batchRequest struct {
-	args []any
-	query string
-}
-
-func drainBatch(ctx context.Context, txn *stdsql.Tx, query string, batch []any, argCount int, readerSuffix string) error {
-	var buff bytes.Buffer
-	var writer = csv.NewWriter(&buff)
-
-	t := time.Now()
-	var stringBatch = make([]string, len(batch))
-	for i, v := range batch {
+func rowToCSVRecord(row []any) ([]string, error) {
+	var record = make([]string, len(row))
+	for i, v := range row {
 		switch value := v.(type) {
 			case []byte:
-				stringBatch[i] = string(value)
+				record[i] = string(value)
 			case string:
-				stringBatch[i] = value
+				record[i] = value
 			case nil:
 				// See https://dev.mysql.com/doc/refman/8.0/en/problems-with-null.html
-				stringBatch[i] = "NULL"
+				record[i] = "NULL"
 			case bool:
 				if value == false {
-					stringBatch[i] = "0"
+					record[i] = "0"
 				} else {
-					stringBatch[i] = "1"
+					record[i] = "1"
 				}
 			default:
 				b, err := json.Marshal(value)
 				if err != nil {
-					return fmt.Errorf("encoding value as json: %w", err)
+					return nil, fmt.Errorf("encoding value as json: %w", err)
 				}
-				stringBatch[i] = string(b)
+				record[i] = string(b)
 		}
 	}
 
-	for i := 0; i < len(stringBatch); i += argCount {
-		if err := writer.Write(stringBatch[i:(i+argCount)]); err != nil {
-			return fmt.Errorf("writing csv record: %w", err)
-		}
+	return record, nil
+}
+
+type batchMeta struct {
+	// use w to write csv records for a batch
+	w *csv.Writer
+	// use buff to check how much data has been written so far
+	buff *bytes.Buffer
+}
+
+func (batch batchMeta) Write(converted []any) error {
+	record, err := rowToCSVRecord(converted)
+	if err != nil {
+		return fmt.Errorf("error encoding row to CSV: %w", err)
 	}
 
-	writer.Flush()
-	if err := writer.Error(); err != nil {
-		return fmt.Errorf("flush writing csv: %w", err)
+	batch.w.Write(record)
+	batch.w.Flush()
```

Comment:
Do we need to flush after every write? Or is it sufficient to flush in drain batch?